### PR TITLE
Fix Windows-only NullReferenceException and add more ModuleInstaller tests

### PR DESCRIPTION
**Pull request by ayan4m1**
If the user attempts to uninstall a mod in Windows, there is a NullReferenceException where AddParentDirectories tries to compare a null to a string. It is also continuing to walk up above the GameData directory, adding my Program Files directory to the list of directories to delete.

The new code normalizes all compared paths and checks for several edge cases, improving its reliability.

I added new tests to ensure the Windows path oddities I saw don't creep back in.

### CODE REVIEW

File: Core/ModuleInstaller.cs [hunk 1]
```diff
@@ -890,28 +890,39 @@ private void Uninstall(string modName)
...
         {
             var newDirectories = new HashSet<string>();
-            foreach (string directory in directories)
+            if (directories == null)
             {
-                if ( directory.Contains(ksp.GameData()))
+                return newDirectories;
+            }
+
+            var normalGameData = KSPPathUtils.NormalizePath(ksp.GameData());
```
Comment (politas) [inline Core/ModuleInstaller.cs]: ksp.GameData() is already normalized 

        public string GameData()
        {
            return KSPPathUtils.NormalizePath(
                Path.Combine(GameDir(), "GameData")
            );
        }

File: Core/ModuleInstaller.cs [hunk 2]
```diff
@@ -890,28 +890,39 @@ private void Uninstall(string modName)
...
+            foreach (var directory in directories)
+            {
+                if (!directory.Contains(normalGameData))
```
Comment (politas) [inline Core/ModuleInstaller.cs]: At this point directory has not been normalized

File: Core/ModuleInstaller.cs [hunk 3]
```diff
@@ -890,28 +890,39 @@ private void Uninstall(string modName)
...
+                    newDirectories.Add(KSPPathUtils.NormalizePath(directory));
                 }
-                else
+
+                var path = KSPPathUtils.NormalizePath(directory);
+                while ((path != normalGameData) && !newDirectories.Contains(path))
                 {
-                    newDirectories.Add(directory);
+                    newDirectories.Add(path);
+                    var parent = Directory.GetParent(path);
```
Comment (politas) [inline Core/ModuleInstaller.cs]: We should probably use 'KSPPathUtils.GetLeadingPathElements(path)` here, so we'd then be checking for an empty string next

File: Core/ModuleInstaller.cs [hunk 4]
```diff
@@ -889,29 +889,41 @@ private void Uninstall(string modName)
...
+
+            var gameData = KSPPathUtils.NormalizePath(ksp.GameData());
+            return directories
+                .Where(dir => !string.IsNullOrWhiteSpace(dir))
+                // normalize before we deduplicate this list
+                .Select(KSPPathUtils.NormalizePath)
+                // remove any duplicate paths
+                .Distinct()
+                // get rid of paths that are above GameData
+                .Where(dir => dir.Length > gameData.Length)
```
Comment (politas) [inline Core/ModuleInstaller.cs]: This would exclude any directory with a name shorter than 9 characters that was installed to Gameroot?

File: Core/ModuleInstaller.cs [hunk 5]
```diff
@@ -889,29 +889,41 @@ private void Uninstall(string modName)
...
         {
-            var newDirectories = new HashSet<string>();
-            foreach (string directory in directories)
+            if (directories == null || directories.Count == 0)
             {
-                if ( directory.Contains(ksp.GameData()))
+                return new HashSet<string>();
+            }
+
+            var gameData = KSPPathUtils.NormalizePath(ksp.GameData());
```
Comment (politas) [inline Core/ModuleInstaller.cs]: This becomes KSPPathUtils.NormalizePath(KSPPathUtils.NormalizePath( Path.Combine(GameDir(), "GameData")))

File: Core/ModuleInstaller.cs [hunk 6]
```diff
@@ -889,29 +889,41 @@ private void Uninstall(string modName)
...
+                    var relativeHead = KSPPathUtils.ToRelative(dir, gameData);
+                    var pathArray = relativeHead.Split('/');
+                    var builtPath = string.Empty;
+                    foreach (var path in pathArray)
                     {
-                        newDirectories.Add(path);
-                        DirectoryInfo parent = Directory.GetParent(path);
-                        path = parent.FullName;
+                        builtPath += path + '/';
+                        results.Add(KSPPathUtils.ToAbsolute(builtPath, gameData));
```
Comment (politas) [inline Core/ModuleInstaller.cs]: Things can be installed to other directories. We should include any directory not under gameData without iterating for parents.

**Comment by politas** [comment]
TestNullPath() is failing. It looks like `_installer.AddParentDirectories(new HashSet<string>() { sysRoot });` is getting past the sysroot test, at which point ToAbsolute throws a kraken due to sysroot already being absolute.
